infra/Pos/Slotting/Types.hs

```diff
@@ -23,14 +33,65 @@ data EpochSlottingData = EpochSlottingData
 instance NFData EpochSlottingData
 
 -- | Data necessary for slotting to work which is basically part of GState.
-data SlottingData = SlottingData
-    { sdPenult      :: !EpochSlottingData
-    -- ^ 'EpochSlottingData' for penult (i. e. 'sdPenultEpoch') epoch.
-    , sdLast        :: !EpochSlottingData
-    -- ^ 'EpochSlottingData' for last epoch.
-    -- Last means 'sdPenultEpoch + 1'.
-    , sdPenultEpoch :: !EpochIndex
-    -- ^ Penult epoch for which 'EpochSlottingData' is known.
-    } deriving (Eq, Show, Generic)
+-- TODO: Reflect the fact that this will always have 2+ elements
+-- TODO: Make this a newtype wrapper
+type SlottingData = HashMap EpochIndex EpochSlottingData
+
+-- PRIVATE
+-- AJ: TODO: GOOD DEFAULT VALUE FOR SLOT DURATION?
+ensureEpochSlottingDataExists :: Maybe EpochSlottingData -> EpochSlottingData
+ensureEpochSlottingDataExists = fromMaybe $ EpochSlottingData 0 0
+
+-- PRIVATE
+-- AJ: TODO: DELETE
+-- ensureEpochIndexExists :: Maybe EpochIndex -> EpochIndex
+-- ensureEpochIndexExists = fromMaybe 0
+
+-- | Get the latest epoch index.
+getLastEpochIndex :: SlottingData -> Maybe EpochIndex
+getLastEpochIndex = fmap maximum . nonEmpty . HM.keys
```

Comment:
It's inefficient and this function is used quite often, so I propose two approaches, choose what you like more (I already wrote about it below):
1. Use ordered `Map` from `containers`.
2. Store last epoch explicitly.